node/src/components/chainspec_loader.rs

```diff
@@ -298,6 +425,224 @@ impl ChainspecLoader {
+    /// a proposer and doesn't contain any deploys or transfers. It is the only block in its era,
+    /// and no consensus instance is run for era 0 or an upgrade point era.
+    fn execute_immediate_switch_block<REv>(
+        &self,
+        effect_builder: EffectBuilder<REv>,
+        maybe_previous_block_header: Option<Box<BlockHeader>>,
+        initial_pre_state: ExecutionPreState,
+        finalized_block: FinalizedBlock,
+    ) -> Effects<Event>
+    where
+        REv: From<ContractRuntimeRequest>
+            + From<StorageRequest>
+            + From<MarkBlockCompletedRequest>
+            + Send,
+    {
+        let protocol_version = self.chainspec.protocol_version();
+        async move {
+            let block_and_execution_effects = effect_builder
+                .execute_finalized_block(
+                    protocol_version,
+                    initial_pre_state,
+                    finalized_block,
+                    vec![],
+                    vec![],
+                )
+                .await?;
+            // We need to store the block now so that the era supervisor can be properly
+            // initialized in the participating reactor's constructor.
+            effect_builder
+                .put_block_to_storage(block_and_execution_effects.block.clone())
+                .await;
+            effect_builder
+                .mark_block_completed(block_and_execution_effects.block.height())
+                .await;
+            info!(
+                immediate_switch_block = ?block_and_execution_effects.block.clone(),
+                "immediate switch block after upgrade/genesis stored"
+            );
+            Ok(block_and_execution_effects)
+        }
+        .event(move |result| Event::ExecuteImmediateSwitchBlockResult {
+            maybe_previous_block_header,
+            result,
+        })
+    }
+
+    fn handle_execute_immediate_switch_block_result<REv>(
+        &mut self,
+        effect_builder: EffectBuilder<REv>,
+        maybe_previous_block_header: Option<Box<BlockHeader>>,
+        result: Result<BlockAndExecutionEffects, BlockExecutionError>,
+    ) -> Effects<Event>
+    where
+        REv: From<ControlAnnouncement> + Send,
+    {
+        let immediate_switch_block_and_exec_effects = match result {
+            Ok(block_and_execution_effects) => block_and_execution_effects,
+            Err(error) => {
+                error!(%error, "failed to execute block");
+                return fatal!(effect_builder, "{}", error).ignore();
+            }
+        };
+
+        // If the switch block before the immediate switch block is `None`, we use the
+        // `next_era_validators` of the immediate switch block to sign it.  This is the case at
+        // genesis and for an emergency upgrade.
+        let maybe_era_end = maybe_previous_block_header
+            .as_deref()
+            .unwrap_or_else(|| immediate_switch_block_and_exec_effects.block.header())
+            .era_end();
```

Comment:
Yes. I actually mentioned this issue on Slack. A part of this will have to be reworked in order to correctly read the validators that are supposed to sign the immediate switch block.

Reply:
Changed in [89d9a3c](https://github.com/casper-network/casper-node/pull/3280/commits/89d9a3c93079055e6a38ee27d75a9f6a2102cc60)